src/libexpr/include/nix/expr/parser-state.hh

```diff
@@ -283,54 +315,64 @@ ParserState::stripIndentation(const PosIdx pos, std::vector<std::pair<PosIdx, st
             }
         }
     }
+    if (atStartOfLine)
+        finalBlankLine = curIndent;
 
     /* Strip spaces from each line. */
     auto * es2 = new std::vector<std::pair<PosIdx, Expr *>>;
     atStartOfLine = true;
     size_t curDropped = 0;
-    size_t n = es.size();
+    size_t n = 0;
     auto i = es.begin();
     const auto trimExpr = [&](Expr * e) {
         atStartOfLine = false;
         curDropped = 0;
         es2->emplace_back(i->first, e);
     };
     const auto trimString = [&](const StringToken & t) {
-        std::string s2;
-        for (size_t j = 0; j < t.l; ++j) {
+        auto finalLineTrim = n == es.size() - 1 ? finalBlankLine : 0;
+        /* try to pre-calculate exactly how big of a string we need. In weird
+         * rare cases we can't efficiently pre-calculate it and will end up
+         * over-allocating. See comment above.
+         */
+        size_t size = 1 + t.l - nrIndentedLines[n] * minIndent - finalLineTrim;
+        if (size == 1) // ignore empty strings before we allocate
+            return;
+        char * s2 = (char *) alloc.allocate(size);
+        size_t end = t.l - finalLineTrim;
+        size_t c = 0;
+        for (size_t j = 0; j < end; ++j) {
             if (atStartOfLine) {
                 if (t.p[j] == ' ') {
                     if (curDropped++ >= minIndent)
-                        s2 += t.p[j];
+                        s2[c++] = t.p[j];
                 } else if (t.p[j] == '\n') {
                     curDropped = 0;
-                    s2 += t.p[j];
+                    s2[c++] = t.p[j];
                 } else {
                     atStartOfLine = false;
                     curDropped = 0;
-                    s2 += t.p[j];
+                    s2[c++] = t.p[j];
                 }
             } else {
-                s2 += t.p[j];
+                s2[c++] = t.p[j];
                 if (t.p[j] == '\n')
                     atStartOfLine = true;
             }
         }
-
-        /* Remove the last line if it is empty and consists only of
-           spaces. */
-        if (n == 1) {
-            std::string::size_type p = s2.find_last_of('\n');
-            if (p != std::string::npos && s2.find_first_not_of(' ', p + 1) == std::string::npos)
-                s2 = std::string(s2, 0, p + 1);
-        }
-
-        // Ignore empty strings for a minor optimisation and AST simplification
-        if (s2 != "") {
-            es2->emplace_back(i->first, new ExprString(std::move(s2)));
-        }
+        if (perfectPreallocate[n])
+            assert(c == size - 1);
+        else
+            assert(c < size);
+        // We should have caught empty strings before allocation. The only case
+        // in which we have to over-allocate is a case with an empty line, which
+        // is therefore not empty.
+        assert(c > 0);
+        s2[c] = '\0';
+
+        es2->emplace_back(i->first, new ExprString(s2));
```

Comment:
I'm wondering whether it's possible to allocate the AST nodes with the buffer allocator.

Reply:
I think that is supposed to come next? See the main issue.

Reply:
I'm intending to move the AST nodes into big vectors. How exactly that relates to the buffer allocator isn't clear to me yet since they have to be able to move to accommodate growth.

But indeed I haven't forgotten to consider where the `Expr`s are going, I'm just trying to do things one small step at a time.